File: ios-app/AnagramArena/Sources/Core/Engine/WordConstructability.swift
import Foundation

struct LetterFrequency {
    private(set) var counts: [Character: Int] = [:]

    init(_ letters: [Character]) {
        for char in letters {
            let normalized = Character(String(char).uppercased())
            counts[normalized, default: 0] += 1
        }
    }

    mutating func consume(_ char: Character) -> Bool {
        let normalized = Character(String(char).uppercased())
        guard let current = counts[normalized], current > 0 else {
            return false
        }

        counts[normalized] = current - 1
        return true
    }
}

struct WordConstructabilityChecker {
    func canConstruct(word: String, from letters: [Character]) -> Bool {
        let normalizedWord = WordNormalizer.normalize(word)
        guard !normalizedWord.isEmpty else { return false }

        var frequency = LetterFrequency(letters)

        for char in normalizedWord {
            if !frequency.consume(char) {
                return false
            }
        }

        return true
    }
}

enum WordNormalizer {
    static func normalize(_ input: String) -> String {
        input
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
    }

    static func isAlphabetical(_ input: String) -> Bool {
        !input.isEmpty && input.unicodeScalars.allSatisfy(CharacterSet.letters.contains)
    }
}

File: android-app/core/src/main/kotlin/com/bhan796/anagram/core/validation/WordConstructability.kt
package com.bhan796.anagram.core.validation

class LetterFrequency(letters: List<Char>) {
    private val counts: MutableMap<Char, Int> = mutableMapOf()

    init {
        for (char in letters) {
            val normalized = char.uppercaseChar()
            counts[normalized] = (counts[normalized] ?: 0) + 1
        }
    }

    fun consume(char: Char): Boolean {
        val normalized = char.uppercaseChar()
        val current = counts[normalized] ?: return false
        if (current <= 0) return false

        counts[normalized] = current - 1
        return true
    }
}

class WordConstructabilityChecker {
    fun canConstruct(word: String, from: List<Char>): Boolean {
        val normalizedWord = WordNormalizer.normalize(word)
        if (normalizedWord.isEmpty()) return false

        val frequency = LetterFrequency(from)
        for (char in normalizedWord) {
            if (!frequency.consume(char)) {
                return false
            }
        }

        return true
    }
}

object WordNormalizer {
    fun normalize(input: String): String = input.trim().lowercase()

    fun isAlphabetical(input: String): Boolean = input.isNotEmpty() && input.all { it.isLetter() }
}